File: src/test/java/org/frc1778/freezylib/util/MeasurementTest.java
package org.frc1778.freezylib.util;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

import org.assertj.core.data.Offset;
import org.junit.jupiter.api.Test;

public class MeasurementTest {

  @Test
  public void testConversion() {
    var kilometer = new Measurement(1.5, Measurement.Length.KILOMETERS);
    assertThat(kilometer.convertTo(Measurement.Length.INCHES).getValue())
        .isEqualTo(59055.118110236224, Offset.offset(1e-9));

    var fahrenheit = new Measurement(212, Measurement.Temperature.DEGREE_FAHRENHEIT);
    assertThat(fahrenheit.convertTo(Measurement.Temperature.DEGREE_CELSIUS).getValue())
        .isEqualTo(100.0, Offset.offset(1e-9));

    var pounds = new Measurement(1.5, Measurement.Mass.POUNDS);
    assertThat(pounds.convertTo(Measurement.Mass.GRAMS).getValue())
        .isEqualTo(680.3885564999999, Offset.offset(1e-9));
  }

  @Test
  public void incompatibleTypesShouldThrowException() {
    assertThatExceptionOfType(IllegalArgumentException.class)
        .isThrownBy(
            () ->
                new Measurement(0.0, Measurement.Acceleration.GRAVITY)
                    .convertTo(Measurement.Duration.HOURS));
  }
}

File: src/test/kotlin/org/frc1778/freezylib/util/MeasurementTest.kt
package org.frc1778.freezylib.util

import io.kotlintest.data.forall
import io.kotlintest.matchers.doubles.plusOrMinus
import io.kotlintest.shouldBe
import io.kotlintest.shouldThrow
import io.kotlintest.specs.StringSpec
import io.kotlintest.tables.row
import java.lang.IllegalArgumentException

class MeasurementTest : StringSpec() {

    init {
        "An incompatible type should throw an IllegalArgumentException" {
            shouldThrow<IllegalArgumentException> { Measurement(0.0, Measurement.Acceleration.GRAVITY).convertTo(Measurement.Duration.HOURS) }
        }

        "Conversion tests" {
            forall(
                    row(1.5, Measurement.Length.KILOMETERS, Measurement.Length.FEET, 4921.26),
                    row(212.0, Measurement.Temperature.DEGREE_FAHRENHEIT, Measurement.Temperature.DEGREE_CELSIUS, 100.0),
                    row(1.5, Measurement.Mass.POUNDS, Measurement.Mass.GRAMS, 680.389)
            ) { input: Double, inputUnit: Measurement.Base, convertedUnit: Measurement.Base, converted: Double ->
                Measurement(input, inputUnit).convertTo(convertedUnit).value shouldBe (converted plusOrMinus 1e-3)
            }
        }
    }
}
